File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.java
class Solution {
    public int shortestPathBinaryMatrix(int[][] grid) {
        if (grid[0][0] != 0 || grid[grid.length - 1][grid.length - 1] != 0) {
            return -1;
        }

        int n = grid.length;
        Queue<int[]> q = new LinkedList<>();
        int[][] dist = new int[n][n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                dist[i][j] = Integer.MAX_VALUE;
            }
        }

        dist[0][0] = 1; // Starting point has distance 1
        q.add(new int[]{0, 0, 1}); // {row, col, distance}

        int[] dr = {-1, -1, 0, 1, 1, 1, 0, -1};
        int[] dc = {0, 1, 1, 1, 0, -1, -1, -1};

        while (!q.isEmpty()) {
            int[] tmp = q.poll();
            int row = tmp[0];
            int col = tmp[1];
            int dis = tmp[2];

            if (row == n - 1 && col == n - 1) {
                return dis; // Reached destination
            }

            for (int i = 0; i < 8; i++) {
                int nrow = row + dr[i];
                int ncol = col + dc[i];

                if (nrow >= 0 && nrow < n && ncol >= 0 && ncol < n && grid[nrow][ncol] == 0 && dis + 1 < dist[nrow][ncol]) {
                    dist[nrow][ncol] = dis + 1;
                    q.add(new int[]{nrow, ncol, dis + 1});
                }
            }
        }
        return -1; // No path found
    }
    
}

File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.py
from collections import deque

class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        n = len(grid)
        
        # talk: start or end blocked
        if grid[0][0] != 0 or grid[n-1][n-1] != 0:
            return -1
        
        # 8 directions
        directions = [(-1,-1), (-1,0), (-1,1),
                    (0,-1),         (0,1),
                    (1,-1), (1,0),  (1,1)]
        
        # queue for BFS with (row, col, path length)
        queue = deque()
        queue.append( (0,0,1) )
        
        # mark visited
        grid[0][0] = 1
        
        while queue:
            row, col, length = queue.popleft()
            
            # talk: reached target
            if row == n-1 and col == n-1:
                return length
            
            # explore 8 neighbors
            for dr, dc in directions:
                r, c = row + dr, col + dc
                if 0 <= r < n and 0 <= c < n and grid[r][c] == 0:
                    queue.append( (r, c, length+1) )
                    grid[r][c] = 1  # mark visited
                    
        return -1

            